presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java

```diff
@@ -296,66 +293,53 @@ private void loadPartition(HivePartitionMetadata partition)
         }
 
         FileSystem fs = hdfsEnvironment.getFileSystem(path);
+        HiveFileIterator iterator = new HiveFileIterator(path, fs, directoryLister, namenodeStats, partitionName, inputFormat, schema, partitionKeys, effectivePredicate);
         if (bucket.isPresent()) {
-            Optional<FileStatus> bucketFile = getBucketFile(bucket.get(), fs, path);
-            if (bucketFile.isPresent()) {
-                FileStatus file = bucketFile.get();
-                BlockLocation[] blockLocations = fs.getFileBlockLocations(file, 0, file.getLen());
-                boolean splittable = isSplittable(inputFormat, fs, file.getPath());
+            List<LocatedFileStatus> locatedFileStatuses = listAndSortBucketFiles(iterator, bucket.get().getBucketCount());
+            FileStatus file = locatedFileStatuses.get(bucket.get().getBucketNumber());
+            BlockLocation[] blockLocations = fs.getFileBlockLocations(file, 0, file.getLen());
+            boolean splittable = isSplittable(inputFormat, fs, file.getPath());
 
-                hiveSplitSource.addToQueue(createHiveSplits(
-                        partitionName,
-                        file.getPath().toString(),
-                        blockLocations,
-                        0,
-                        file.getLen(),
-                        schema,
-                        partitionKeys,
-                        splittable,
-                        session,
-                        effectivePredicate));
-                return;
-            }
+            hiveSplitSource.addToQueue(createHiveSplits(
+                    partitionName,
+                    file.getPath().toString(),
+                    blockLocations,
+                    0,
+                    file.getLen(),
+                    schema,
+                    partitionKeys,
+                    splittable,
+                    session,
+                    effectivePredicate));
+            return;
         }
 
-        HiveFileIterator iterator = new HiveFileIterator(path, fs, directoryLister, namenodeStats, partitionName, inputFormat, schema, partitionKeys, effectivePredicate);
         fileIterators.addLast(iterator);
     }
 
-    private static Optional<FileStatus> getBucketFile(HiveBucket bucket, FileSystem fs, Path path)
+    private static List<LocatedFileStatus> listAndSortBucketFiles(HiveFileIterator hiveFileIterator, int bucketCount)
     {
-        FileStatus[] statuses = listStatus(fs, path);
+        ArrayList<LocatedFileStatus> list = new ArrayList<>(bucketCount);
 
-        if (statuses.length != bucket.getBucketCount()) {
-            return Optional.empty();
-        }
-
-        Map<String, FileStatus> map = new HashMap<>();
-        List<String> paths = new ArrayList<>();
-        for (FileStatus status : statuses) {
-            if (!isFile(status)) {
-                return Optional.empty();
+        while (hiveFileIterator.hasNext()) {
+            if (list.size() >= bucketCount) {
```

Comment:
Place this check after the directory check, so you get the directory message if the last file is a directory.